client/client-multi/client-c/src/kaa_logging.c

```diff
@@ -220,37 +218,51 @@ kaa_error_t kaa_logging_request_serialize(kaa_log_collector_t *self, kaa_platfor
+            ? self->max_log_bucket_size
+            : tmp_writer.end - tmp_writer.current;
+    KAA_LOG_TRACE(self->logger, KAA_ERR_NONE, "Extracting log records... (remaining bucket size is %d)", remaining_size);
 
     uint16_t records_count = 0;
-    kaa_log_entry_t entry = self->log_storage.get_next_record(self->log_storage.context
-            , self->log_bucket_id, MAX_RECORD_SIZE(remaining_size));
-    for (; entry.record_data; entry = self->log_storage.get_next_record(self->log_storage.context, self->log_bucket_id, MAX_RECORD_SIZE(remaining_size))) {
-        KAA_LOG_TRACE(self->logger, KAA_ERR_NONE, "Got record {%p}, size: %zu", entry.record_data, entry.record_size);
-        ++records_count;
-        remaining_size -= (kaa_aligned_size_get(entry.record_size) + sizeof(uint32_t));
-
-        *((uint32_t *) writer->current) = KAA_HTONL(entry.record_size);
-        writer->current += sizeof(uint32_t);
-
-        if (kaa_platform_message_write_aligned(writer, entry.record_data, entry.record_size)) {
-            KAA_LOG_ERROR(self->logger, KAA_ERR_WRITE_FAILED, "Failed to write the log record buffer {%p}", entry.record_data);
-            if (self->log_storage.upload_failed)
-                self->log_storage.upload_failed(self->log_storage.context, self->log_bucket_id);
-            return KAA_ERR_WRITE_FAILED;
+
+    while (!error) {
+        size_t record_len = 0;
+        error = ext_log_storage_write_next_record(self->log_storage_context
+                , tmp_writer.current + sizeof(uint32_t), remaining_size - sizeof(uint32_t)
+                , self->log_bucket_id, &record_len);
+        switch (error) {
+        case KAA_ERR_NONE:
+            ++records_count;
+            *((uint32_t *) tmp_writer.current) = KAA_HTONL(record_len);
+            tmp_writer.current += (sizeof(uint32_t) + record_len);
+            kaa_platform_message_write_alignment(&tmp_writer);
+            remaining_size -= (kaa_aligned_size_get(record_len) + sizeof(uint32_t));
+            break;
+        case KAA_ERR_NOT_FOUND:
+        case KAA_ERR_INSUFFICIENT_BUFFER:
+            // These errors are normal if they appear after at least one record got serialized
+            if (!records_count) {
+                KAA_LOG_ERROR(self->logger, error, "Failed to write the log record");
+                return error;
+            }
+            break;
+        default:
+            KAA_LOG_ERROR(self->logger, error, "Failed to write the log record");
+            return error;
         }
     }
-    total_size += (self->log_properties.max_log_bucket_size - remaining_size);
-    KAA_LOG_TRACE(self->logger, KAA_ERR_NONE, "Extracted log records. Total records count = %u. Total extension size = %lu", records_count, total_size);
+
+    size_t total_size = tmp_writer.current - writer->current - KAA_EXTENSION_HEADER_SIZE;
+    KAA_LOG_TRACE(self->logger, KAA_ERR_NONE, "Extracted %u log records; total extension size %lu", records_count, total_size);
```

Comment:
1. Please, use %zu specifier for the second argument.
2. Maybe better to replace "total extension size" with "payload size"?